### ISSUE TITLE
Create an authentication system

### ISSUE THREAD

**Issue by raulcontrerasrubio** [quienmedebe/api-rest#24]
A client should be able to create an account, log in to their account, and reset their password if for some reason they forget it.

The authentication with the server from the client side could be achieved using [JWT](https://jwt.io/).

**Comment by raulcontrerasrubio** [comment]
A number of requirements are needed to implement this feature. Also, we must provide some limitation to the configuration options.

- The provider to log in an account must be decoupled from the user account inside the application.
- The user should be able to log into their account from multiple providers.
- The initial implementation of this feature will only support email + password authentication.

Email and password authentication requirements:
[ ] A user can sign up into the application. The endpoint should support extra properties for future account management.
[ ] A user can log in into the application using their email and password
[ ] A user can reset their password using their email. There must be an endpoint to send an email to that email and another endpoint to reset their password using the token sent in the email.
[ ] There must be an endpoint to get the token needed to create requests to protected endpoints.
[ ] There must be an endpoint to refresh the token when the expire date is near or the token has been compromised.
[ ] These endpoints must be protected using a special rate limiter

To add tests to this system, the creation of User factories will be useful.

The authentication module must have a provider table to save its data. Maybe the use of Sequelize could be appropriate. As a DB engine, we can use MySQL or Postgres. The database service will be needed to address other issues. That module is blocking, so it has to be created here.

Another solution is to use a specific database to manage providers, but I think it would be overengineering.

Inside this issue, we will not create an account module. That will be addressed in another issue. The JWT will return the provider used and nothing else at the moment.

Do we get the `req.user` property inside a route with JWT? If it is true, how are we going to standardize the `req.user` property? Maybe that is a concern of a future account management module.

Future possible implementations:
- [Two Factor Authentication](https://www.npmjs.com/package/node-2fa)
- [Anonymous authentication](https://www.npmjs.com/package/passport-anonymous)
- Social providers: [Google](https://github.com/jaredhanson/passport-google-oauth2), [Facebook](https://github.com/jaredhanson/passport-facebook)

**Comment by raulcontrerasrubio** [comment]
The provider and the user account must be part of the same service. It is absurd to take a provider without an account. Also, the API gateway is in charge of authenticating the incoming requests.

Resources: [https://microservices.io/patterns/security/access-token.html](https://microservices.io/patterns/security/access-token.html)

**Comment by raulcontrerasrubio** [comment]
After some work on this feature, I have come up with some questions:
- Should the signup route return an access_token property on success? It depends. Does the account need to be activated before making any request? In that case, I suppose no; however, maybe the access_token could be returned anyways to use it whenever the account is activated or the activation of the account could be delayed showing a warning on the dashboard of the user. One could manage the activation of the account within the token to provide more flexibility.
- How am I going to implement a refresh token or, how am I going to handle an expired token?
- Is the reset password functionality exclusive for the email provider? Yes
- Is the activation of the account exclusive for the email provider? Yes
- Is the database module external to the application? I have some trouble with the location of the database management, Redis client, logger, error management...
- How am I going to manage the tests? It seems the different folder of the tests is not appropriate. How am I going to make the unit tests? Should the integration tests run across a real database? The rate limiter is part of the API. Is the Redis instance needed in those integration tests?

This module is a little more complex than I wanted to, but I am not surprised about it.

**Comment by raulcontrerasrubio** [comment]
Update of this issue:

I have implemented a JWT system with access and refresh tokens. These tokens are received after login, signup, and refresh the current token.

Speaking of the test structure, I think I have stabilized them. To implement a feature it is more important to create the integration tests and later implement the unit tests because the individual functions are likely to change before creating a stable API.
